File: services/api/app/schemas/referrals.py
from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class StrictSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class DispatchStatus(StrEnum):
    DRAFT = "draft"
    AWAITING_APPROVAL = "awaiting_approval"
    APPROVED = "approved"
    QUEUED = "queued"
    SENDING = "sending"
    SENT = "sent"
    DELIVERY_UNKNOWN = "delivery_unknown"
    FAILED = "failed"
    CANCELLED = "cancelled"


class RegistrationStatus(StrEnum):
    UNVERIFIED = "unverified"
    PENDING = "pending"
    REGISTERED = "registered"
    REJECTED = "rejected"


class HandlingStatus(StrEnum):
    UNASSIGNED = "unassigned"
    AWAITING_ACCEPTANCE = "awaiting_acceptance"
    ACCEPTED = "accepted"
    IN_PROGRESS = "in_progress"
    DECLINED = "declined"
    COMPLETED = "completed"


class RoutingOption(StrictSchema):
    channel_id: UUID
    target_unit_id: UUID
    target_name: str
    channel_name: str
    mode: str
    authority_source: dict[str, Any]
    is_simulated: bool


class RoutingOptionsResponse(StrictSchema):
    report_id: UUID
    items: list[RoutingOption]
    needs_review: bool


class ReferralPackageInput(StrictSchema):
    summary: str = Field(min_length=1, max_length=2000)
    chronology: str = Field(min_length=1, max_length=5000)
    requested_action: str = Field(min_length=1, max_length=2000)
    attachment_ids: list[UUID] = Field(default_factory=list, max_length=10)
    share_citizen_identity: bool = False

    @field_validator("summary", "chronology", "requested_action")
    @classmethod
    def normalize_text(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("value must not be blank")
        return normalized

    @field_validator("attachment_ids")
    @classmethod
    def require_unique_attachments(cls, value: list[UUID]) -> list[UUID]:
        if len(value) != len(set(value)):
            raise ValueError("attachment_ids must be unique")
        return value


class ReferralCreate(StrictSchema):
    channel_id: UUID
    request_key: UUID
    package: ReferralPackageInput


class ReferralApproval(StrictSchema):
    package_version: int = Field(gt=0)
    package_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    reason: str = Field(min_length=1, max_length=1000)

    @field_validator("reason")
    @classmethod
    def normalize_reason(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("reason must not be blank")
        return normalized


class ReferralDispatch(StrictSchema):
    operation_key: UUID


class ReferralCancel(StrictSchema):
    reason: str = Field(min_length=1, max_length=1000)

    @field_validator("reason")
    @classmethod
    def normalize_reason(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("reason must not be blank")
        return normalized


class ReferralPackageSnapshot(StrictSchema):
    summary: str
    chronology: str
    requested_action: str
    attachment_count: int = Field(ge=0)
    share_citizen_identity: bool


class ReferralTask(StrictSchema):
    id: UUID
    referral_id: UUID
    task_type: str
    status: str
    assigned: bool
    next_action: str
    due_at: datetime | None
    blocked_reason: str | None
    created_at: datetime
    updated_at: datetime


class ReferralProgress(StrictSchema):
    id: UUID
    report_id: UUID
    source_unit_id: UUID
    channel_id: UUID
    target_name: str
    channel_name: str
    active_package_version: int
    package_hash: str
    dispatch_status: DispatchStatus
    registration_status: RegistrationStatus
    handling_status: HandlingStatus
    external_reference: str | None
    evidence_reference: str | None
    is_simulated: bool
    package_snapshot: ReferralPackageSnapshot | None = None
    next_action: str | None = None
    updated_at: datetime


class ReferralDispatchResponse(StrictSchema):
    referral: ReferralProgress
    operation_key: UUID
    job_status: str
    replayed: bool


class ReferralWorkerResult(StrictSchema):
    processed: bool
    job_id: UUID | None = None
    dispatch_status: DispatchStatus | None = None

File: apps/dashboard/lib/referrals.ts
import { apiFetch } from "./admin";

export type ReferralDispatchStatus =
  | "draft"
  | "awaiting_approval"
  | "approved"
  | "queued"
  | "sending"
  | "sent"
  | "delivery_unknown"
  | "failed"
  | "cancelled";

export type ReferralRegistrationStatus =
  | "unverified"
  | "pending"
  | "registered"
  | "rejected";

export type ReferralHandlingStatus =
  | "unassigned"
  | "awaiting_acceptance"
  | "accepted"
  | "in_progress"
  | "declined"
  | "completed";

export type ReferralProgress = {
  id: string;
  report_id: string;
  source_unit_id: string;
  channel_id: string;
  target_name: string;
  channel_name: string;
  active_package_version: number;
  package_hash: string;
  dispatch_status: ReferralDispatchStatus;
  registration_status: ReferralRegistrationStatus;
  handling_status: ReferralHandlingStatus;
  external_reference: string | null;
  evidence_reference: string | null;
  is_simulated: boolean;
  package_snapshot: {
    summary: string;
    chronology: string;
    requested_action: string;
    attachment_count: number;
    share_citizen_identity: boolean;
  } | null;
  next_action: string | null;
  updated_at: string;
};

export type RoutingOption = {
  channel_id: string;
  target_unit_id: string;
  target_name: string;
  channel_name: string;
  mode: string;
  authority_source: Record<string, unknown>;
  is_simulated: boolean;
};

export type RoutingOptionsResponse = {
  report_id: string;
  items: RoutingOption[];
  needs_review: boolean;
};

export type ReferralTask = {
  id: string;
  referral_id: string;
  task_type: string;
  status: string;
  assigned: boolean;
  assigned_to_me: boolean;
  next_action: string;
  due_at: string | null;
  blocked_reason: string | null;
  created_at: string;
  updated_at: string;
};

export const getReportReferrals = (reportId: string, signal?: AbortSignal) =>
  apiFetch<ReferralProgress[]>(`/api/v1/reports/${encodeURIComponent(reportId)}/referrals`, { signal });

export const getRoutingOptions = (reportId: string, signal?: AbortSignal) =>
  apiFetch<RoutingOptionsResponse>(`/api/v1/reports/${encodeURIComponent(reportId)}/routing-options`, { signal });

export const getReportTasks = (reportId: string, signal?: AbortSignal) =>
  apiFetch<ReferralTask[]>(`/api/v1/reports/${encodeURIComponent(reportId)}/tasks`, { signal });

export const updateReferralTask = (
  taskId: string,
  payload: { action: "claim" | "release" | "complete"; reason?: string },
) => apiFetch<ReferralTask>(`/api/v1/referral-tasks/${encodeURIComponent(taskId)}`, {
  method: "PATCH",
  body: JSON.stringify(payload),
});

export const approveReferral = (
  referralId: string,
  payload: { package_version: number; package_hash: string; reason: string },
) => apiFetch<ReferralProgress>(`/api/v1/referrals/${encodeURIComponent(referralId)}/approve`, {
  method: "POST",
  body: JSON.stringify(payload),
});

export const dispatchReferral = (referralId: string, operation_key: string) =>
  apiFetch<{ referral: ReferralProgress; operation_key: string; job_status: string; replayed: boolean }>(
    `/api/v1/referrals/${encodeURIComponent(referralId)}/dispatch`,
    { method: "POST", body: JSON.stringify({ operation_key }) },
  );

export const reconcileReferral = (referralId: string) =>
  apiFetch<ReferralProgress>(`/api/v1/referrals/${encodeURIComponent(referralId)}/reconcile`, { method: "POST" });
